### PR TITLE
[python-package] Allow to pass Arrow table as training data

### PR DESCRIPTION
**Pull request by borchero**
# Motivation

This is the first of a set of replacement PRs to simplify https://github.com/microsoft/LightGBM/pull/6022.

# Changes

- Allow to Arrow table to `lgb.Dataset.data`
- Add tests for Arrow C++ implementation and Python binding

### CODE REVIEW

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for starting to break this into smaller pieces! I only had a few minutes to quickly scan this, so left some minor comments. I'll leave a more thorough review when I have time.

Could you please describe here why the approach you're recommending is to use `pyarrow` + converting in that library to corresponding C API functions, instead of having LightGBM link to `libarrow` and using its entrypoints directly?

Linking to `libarrow` might complicate LightGBM's build process and it'd introduce some conversations about how we want to handle packaging, but it'd also mean that Arrow support would also be available to projects integrating directly with LightGBM's C API. It'd also reduce the amount of additional effort needed to support this in the R package, for example.

I'm NOT saying that we definitely should take that approach, I just would like to hear why you chose not to involve linking to `libarrow` here.

File: .gitignore
Comment (borchero) [reply .gitignore:273]: `include/LightGBM/arrow.h` and `include/LightGBM/arrow.tpp` were ignored before adding this. Unsure if this is related to me developing on MacOS? 👀

**Comment by borchero** [comment]
> Could you please describe here why the approach you're recommending is to use pyarrow + converting in that library to corresponding C API functions, instead of having LightGBM link to libarrow and using its entrypoints directly?

My primary thought was that including `libarrow` needlessly blows up the size of the dependencies (in particular, for the Python package where `libarrow` would need to be included). Using the C data interface follows the path that XGBoost has taken to add Arrow support (see https://github.com/dmlc/xgboost/pull/7512). This happens to align with the design choice in other popular software such as Velox or DuckDB.

Since we essentially only need an iterator over C-style arrays, I reckoned that there is no harm in adding this implementation myself.

> Linking to libarrow might complicate LightGBM's build process and it'd introduce some conversations about how we want to handle packaging, but it'd also mean that Arrow support would also be available to projects integrating directly with LightGBM's C API.

That's a fair point, I haven't thought about this tbh. In my mind, LightGBM is only used from Python and R. I don't think I can contribute much to such a conversation though.

---

In any case, I'm happy to update the PR if you come to the conclusion that linking against `libarrow` is preferable :)

File: .gitignore
Comment (jameslamb) [reply .gitignore:273]: Strange! I ran the following to check

```shell
touch include/LightGBM/arrow.h
git check-ignore -v include/LightGBM/arrow.h
```

And got this

```text
.gitignore:271:lightgbm	include/LightGBM/arrow.h
```

That's saying that this line is causing it to be ignored:

https://github.com/microsoft/LightGBM/blob/c60bc739b0a8163e134c3df721b089020f838726/.gitignore#L271

That's there to ignore the LightGBM CLI (which is named `lightgbm`). I'll fix that, sorry!

For this PR, please revert the changes to `.gitignore` and add the files you need to the repo like this:

```shell
git add -f include/LightGBM/arrow.h
git add -f include/LightGBM/arrow.tpp
```

File: include/LightGBM/dataset.h
Comment (jmoralez) [inline include/LightGBM/dataset.h:508]: I think this should be in the PushOneRow function to perform this check per-row instead of per-value.

**Comment by lorentzenchr** [comment]
As pointed out in https://github.com/microsoft/LightGBM/pull/6022#issuecomment-1685255598, [nanoarrow](https://arrow.apache.org/nanoarrow/0.2.0/index.html) might be an alternative.

File: .github/workflows/python_package.yml
Comment (jameslamb) [inline .github/workflows/python_package.yml:6]: please revert all of these indentation and quoting changes, unless you can explain specifically how they're related to the goals of this PR

**Comment by xhochy** [comment]
> I'm NOT saying that we definitely should take that approach, I just would like to hear why you chose not to involve linking to libarrow here.

Linking to `libarrow` is only worth it if you use additional functionality of it that goes beyond passing Arrow structures around. The additional functionality of `libarrow` comes at the cost of dealing with complex dependencies. If you have the luxury of having conda, this is fine but in a `pip`/`CRAN`-only world, it is a big hassle.

File: python-package/lightgbm/arrow.py
Comment (jameslamb) [inline python-package/lightgbm/arrow.py:1]: Could you please move these definitions into `basic.py`, instead of introducing a new top-level module?

Since everything being included here is private (prefixed with `_` and excluded from `__all__`), and therefore never intended to be imported by user code, I don't think there's much benefit to separating it out into a separate module.

File: .github/workflows/python_package.yml
Comment (jameslamb) [inline .github/workflows/python_package.yml:25]: Why was this necessary?

On latest `master`, it's taking only around 20 minutes for all `Python-package` GitHub Actions jobs to complete: https://github.com/microsoft/LightGBM/actions/runs/6097609206

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks @lorentzenchr  and @xhochy for your input, and to @borchero for your thorough answer in https://github.com/microsoft/LightGBM/pull/6034#issuecomment-1676015105.

Let's proceed with the approach already outlined in this PR (no `nanoarrow` or `libarrow). It's the lightest-weight, and I really like that we'd be able to switch to a different approach in the future without breaking users of the Python package, since it doesn't introduce new required dependencies or new code in the public interface (at least of the Python package).

Please see my first small set of additional review comments. We'll try to provide another review in the next few days.

File: .github/workflows/python_package.yml
Comment (borchero) [reply .github/workflows/python_package.yml:25]: Doesn't seem to be, there's another issue that I mistook for a timeout. Will revert this change.

File: .gitignore
Comment (borchero) [reply .gitignore:142]: > CI will run automatically when you push commits here, without needing to wait for a maintainer to click a button

Ah, good thinking :smile: will open another PR!

Comment (jameslamb) [reply .gitignore:142]: ha thanks very much!

File: python-package/lightgbm/arrow.py
Comment (borchero) [reply python-package/lightgbm/arrow.py:1]: Two arguments why I put it in `arrow.py`:

- `basic.py` is already rather unwieldy and I found it rather time-consuming to work with. A smaller module was much easier to handle.
- There are `pyarrow` imports in `arrow.py`: I can't easily put them in `basic.py`, hence, I currently go over `compat.py` (i.e. `basic.py` doesn't import from `.arrow` but rather from `.compat` which, in turn, imports `.arrow` in a `try-except`). I wouldn't know how to put the code in `basic.py` without adding a number of compatibility classes to `compat.py` (which I wouldn't consider particularly ergonomic).

What about naming the file `_arrow.py` to make it clear that it is a private module?

**Comment by borchero** [comment]
Anybody got an idea what's wrong with the `bdist` job? The CI failure message is not particularly helpful and I can't replicate environment creation issues locally :eyes:

File: .ci/test.sh
Comment (jameslamb) [inline .ci/test.sh:127]: This looks like it was a typo on our part. Can you please put up a separate PR fixing it, since it's not specific to these changes and therefore shouldn't have to wait for them to be merged? We'd really appreciate it.

**Comment by jameslamb** [comment]
> Anybody got an idea what's wrong with the `bdist` job? 

In the future when contributing here, please provide a link to the logs when you're asking about a CI job. There are multiple `bdist` CI jobs in this project, and you could save us some time and effort by linking to exactly what you're talking about.

I clicked through the various CI jobs and saw multiple Linux Python jobs failing like this:

```text
tests/c_api_test/test_.py ..                                             [  0%]
...
tests/python_package_test/test_engine.py ..............................s [ 53%]
...s....................................................../__w/1/s/.ci/test.sh: line 201: 10546 Killed                  pytest $BUILD_DIRECTORY/tests
```

([example build](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15170&view=logs&j=b463fb3a-b487-5cfc-fc07-6d216464ba86&t=c16b9cdb-c4c8-59f9-2ae2-b81f62c3a9d0))

Those probably indicate a segmentation fault, invalid memory access, or other process-crashing issue caused by the current state of this branch.

I saw that the Windows `bdist` job is failing like this:

```text
pytest : The term 'pytest' is not recognized as the name of a cmdlet, function, script file, or operable program. 
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At D:\a\1\s\.ci\test_windows.ps1:123 char:1
+ pytest $tests ; Check-Output $?
+ ~~~~~~
    + CategoryInfo          : ObjectNotFound: (pytest:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15170&view=logs&j=ea56812e-e7ae-55d0-6abc-4a217857fa9f&t=e61701f8-4e1b-5007-20cb-a000ab3587fb))

That looks like the type of thing that is temporary and usually resolved by a new CI run. You can trigger a new run by pushing an empty commit.

```shell
git commit --allow-empty -m "empty commit"
git push origin HEAD
```

File: python-package/lightgbm/arrow.py
Comment (jameslamb) [reply python-package/lightgbm/arrow.py:1]: > ... adding a number of compatibility classes to `compat.py` (which I wouldn't consider particularly ergonomic)

This is exactly what I'd like you to do, please.

Based on the current state of this PR, you're only using 1 class (`pyarrow.Table`) and 5 functions (`pyarrow.types.is_floating()`, `pyarrow.types.is_integer()`, `pyarrow.cffi.addressof()`, `pyarrow.cffi.cast()`, `pyarrow.cffi.new()`). That's not so many imports that I think it is worth deviating from the patterns `lightgbm` uses for interacting with external dependencies. There is also "ergonomic" value in similar operations being handled in similar ways throughout the codebase.

> I wouldn't know how to put the code in `basic.py` ...

Put this in `compat.py`:

```python
"""pyarrow"""
try:
    from pyarrow import Table as pa_Table
    from pyarrow.cffi import (
        addressof AS arrow_ffi_addressof,
        cast AS arrow_ffi_cast,
        new AS arrow_ffi_new,
    )
    from pyarrow.types import (
        is_floating as arrow_is_floating,
        is_integer as arrow_is_integer
    )
    PYARROW_INSTALLED = True
except ImportError:
    PYARROW_INSTALLED = False

    class pa_Table:  # type: ignore
        """Dummy class for pa.Table."""

        def __init__(self, *args, **kwargs):
            pass

    arrow_is_integer = None
    arrow_is_floating = None
    arrow_ffi_addressof = None
    arrow_ffi_cast = None
    arrow_ffi_new = None
```

And then in `basic.py`


```python
from .compat import arrow_is_integer, arrow_is_floating, arrow_ffi_addressof, arrow_ffi_cast, arrow_ffi_new

if TYPE_CHECKING:
    from pyarrow.cffi import CData as pyarrow_CData

# ...
class _ArrowCArray:
    """Simple wrapper around the C representation of an Arrow type."""

    n_chunks: int
    chunks: "pyarrow_CData"
    schema: "pyarrow_CData"

    def __init__(self, n_chunks: int, chunks: "pyarrow_CData", schema: "pyarrow_CData"):
        self.n_chunks = n_chunks
        self.chunks = chunks
        self.schema = schema

    @property
    def chunks_ptr(self) -> int:
        """Returns the address of the pointer to the list of chunks making up the array."""
        return int(ffi.cast("uintptr_t", ffi.addressof(self.chunks[0])))

    @property
    def schema_ptr(self) -> int:
        """Returns the address of the pointer to the schema of the array."""
        return int(ffi.cast("uintptr_t", self.schema))
```

If you're not familiar with that type-hints-in-a-string approach in Python, see https://mypy.readthedocs.io/en/stable/runtime_troubles.html.

Comment (borchero) [reply python-package/lightgbm/arrow.py]: Thanks for the elaborate explanation! Done ✅

**Comment by borchero** [comment]
> In the future when contributing here, please provide a link to the logs when you're asking about a CI job

I'm very sorry, I did not realize that there is a bunch more jobs running on Azure, I was only really paying attention to the ones showing up directly on GitHub :eyes: will do in the future!

Do you have an idea why the CI still requires approval @jameslamb? Would be nice to be able to push more often to fix the remaining issue(s) 😬

**Comment by jameslamb** [comment]
> Do you have an idea why the CI still requires approval @jameslamb? Would be nice to be able to push more often to fix the remaining issue(s)

I'm not sure why it's still requiring maintainer approval for your CI runs. Maybe that status gets assigned on GitHub's backend based on the time when you open the PR?

I'll try to keep clicking the button whenever I see notifications here.

**Comment by jameslamb** [comment]
I just updated this with the latest changes from `master`. Once CI builds, I'll give it another review.

**Comment by jameslamb** [comment]
@borchero could you please fix the failing tests? `@` me if you need help.

**Comment by borchero** [comment]
@jameslamb yes, definitely, I will finally take care of this on the weekend! 🙏🏼

**Comment by borchero** [comment]
Sorry it took so long to continue working on this! Everything I can run locally passes now, let's see how the CI is doing 😬

**Comment by borchero** [comment]
Yeah something's still not right with one of the tests... I can't replicate the failure locally (on Apple Silicon using gcc/g++) though. Any guidance how I can properly debug this?

**Comment by borchero** [comment]
Don't think the `R-package` failures are related to this PR

**Comment by borchero** [comment]
The `bdist` Python-package job seems to be failing because the conda env cannot be solved. I cannot yet replicate this locally though (even for `osx-64`)

**Comment by jameslamb** [comment]
> The bdist Python-package job seems to be failing because the conda env cannot be solved

In this project, please be very very specific about what CI job you're talking about, and include a link to the logs. There are multiple CI jobs that could be described as "the `bdist` Python-package job".

For example, the Windows `bdist` job running on Windows is not failing because of a conda env resolution failure.... one of the unit tests is failing.

https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15326&view=logs&j=ea56812e-e7ae-55d0-6abc-4a217857fa9f&t=e61701f8-4e1b-5007-20cb-a000ab3587fb

> FAILED ..\tests\python_package_test\test_engine.py::test_model_size - lightgb...
>
> E           lightgbm.basic.LightGBMError: bad allocation

**Comment by jmoralez** [comment]
Hey, sorry. I think @borchero meant [this one](https://github.com/microsoft/LightGBM/actions/runs/6497230115/job/17648637650?pr=6034) but I reran it.

**Comment by borchero** [comment]
> In this project, please be very very specific about what CI job you're talking about, and include a link to the logs

Yep, sorry, I'm usually referring to the ones listed directly on GitHub and it seemed to uniquely identify [this one](https://github.com/microsoft/LightGBM/actions/runs/6497230115/job/17648637650?pr=6034).

**Comment by borchero** [comment]
Thanks @jmoralez! I don't think re-running will fix the issue though.

> one of the unit tests is failing

@jameslamb that's what I was talking about above, seems to be a unit test that fails across platforms. I'm currently in the process of setting up a Linux instance to be able to debug properly.

**Comment by borchero** [comment]
Can someone help me figure out what the environment looks like that is trying to be created in [this job](https://github.com/microsoft/LightGBM/actions/runs/6497230115/job/17648637650)? I derived the following but I can't even reproduce the failure on macOS GitHub runners (using `micromamba` to solve faster and with better error messages):

```yaml
name: lgb
channels:
  - conda-forge
dependencies:
  - dask-core
  - distributed
  - pandas<2.0
  - python=3.7[build=*cpython]
  - cffi
  - cloudpickle
  - joblib
  - matplotlib
  - numpy
  - psutil
  - pyarrow
  - pytest
  - python-graphviz
  - scikit-learn
  - scipy
```

**Comment by borchero** [comment]
Ha, given that the [Linux logs](https://dev.azure.com/lightgbm-ci/8461a79b-5dce-4085-ad70-4410b7135276/_apis/build/builds/15326/logs/345) have the same "error log" for solving the environment

```
2023-10-12T14:47:44.0781914Z Collecting package metadata (current_repodata.json): ...working... done
2023-10-12T14:48:48.1182869Z Solving environment: ...working... unsuccessful attempt using repodata from current_repodata.json, retrying with next repodata source.
```

It might actually be caused by the timeout. `pyarrow` is quite a huge package, maybe it's the straw that breaks the camel's back.

@jameslamb is there an effort to replace `conda` with something more performant in this repo already (or would you be open to a PR)? Otherwise, I can just increase the timeout here.

**Comment by jameslamb** [comment]
> is there an effort to replace conda with something more performant in this repo already

We already are using `miniforge`:

https://github.com/microsoft/LightGBM/blob/8ed371cee49cf86740b25dd9a4b985a75c9f2dba/.ci/setup.sh#L127-L133

https://github.com/guolinke/lightgbm-ci-docker/blob/9121c2852e1cbbc556f12b4ba70b335989e0d9b4/dockers/manylinux_2_28_x86_64/Dockerfile#L120-L126

At the time that was added (February 2022), we were experiencing issues we couldn't explain that were solved by sticking with `conda` + `conda-forge` instead of `mamba`: https://github.com/microsoft/LightGBM/pull/4953#issuecomment-1030317210

> would you be open to a PR

If you'd like to try switching this project to `mamba` in a new PR, go for it! I'd support it if you can get it working.

It'd also have to be updated in those images we prebuild for some of the Linux CI jobs: https://github.com/guolinke/lightgbm-ci-docker/tree/master/dockers

**Comment by jameslamb** [comment]
> figure out what the environment looks like

Windows:

https://github.com/microsoft/LightGBM/blob/8ed371cee49cf86740b25dd9a4b985a75c9f2dba/.ci/test_windows.ps1#L51-L65

macOS / Linux:

https://github.com/microsoft/LightGBM/blob/8ed371cee49cf86740b25dd9a4b985a75c9f2dba/.ci/test.sh#L125-L142

And please note we use strict channel priority to only get packages from the `conda-forge` channel (via the use of `miniforge`).

**Comment by borchero** [comment]
The culprit in the tests seems to be `test_model_size`. I reckon it causes an OOM but I would be surprised if this is tied to any semantic changes of this PR since the relevant code paths shouldn't be used in this test. @jameslamb do you know whether the memory limit that is used for skipping the test is rather tight already?

**Comment by jameslamb** [comment]
> since the relevant code paths shouldn't be used in this test

I strongly suspect those failing tests are related to changes in this PR.

For example, the `Windows sdist` Azure DevOps job is failing on this PR with that error about a bad allocation in `test_model_size` ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15326&view=logs&j=67ae2ff8-4b33-578b-0813-9c0440091667&t=aed7750d-29a7-50d7-d7f0-fc7145577059)), but the corresponding build is passing on #6140 ([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15328&view=logs&j=67ae2ff8-4b33-578b-0813-9c0440091667&t=aed7750d-29a7-50d7-d7f0-fc7145577059)).

Some possibilities to investigate:

*maybe there memory leaks introduced by the code in this PR, which leave that `test_model_size` test with less free memory available by the time it's run (since the `test_arrow.py` tests run first)
* maybe the addition of `cffi` and `pyarrow` to the test environment leads to some other dependency versions changing which are on the code paths invoked in that test, even if indirectly
    - *you could investigate this by comparing the logs from a build on this PR to the logs from the same CI job on a PR where they're passing, like those from #6410*

**Comment by borchero** [comment]
Thanks for the pointers @jameslamb! Only went through the Linux tests, didn't expect to find something useful in the Windows tests 😬 on it 🙇🏼‍♂️

**Comment by borchero** [comment]
[Disabling arrow tests for debugging apparent OOM]

**Comment by borchero** [comment]
It seems like there's no way around building new container images for Azure: [Linux sdist](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15330&view=logs&j=02a2c3ba-81f8-54e3-0767-5d5adbb0daa9&t=720ee3fa-96d4-5b47-dbf4-01607b74ade2) already chokes during environment creation

**Comment by borchero** [comment]
Ok, latest Azure CI indicates some memory leak IMO, that's a promising avenue.

**Comment by borchero** [comment]
All right, here we go. Memory issues seem to have been a result of crappy test setup rather than an issue with the implementation itself.

**Comment by borchero** [comment]
> It seems like there's no way around building new container images for Azure: [Linux sdist](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=15330&view=logs&j=02a2c3ba-81f8-54e3-0767-5d5adbb0daa9&t=720ee3fa-96d4-5b47-dbf4-01607b74ade2) already chokes during environment creation

This and #6140 are the only open issues regarding CI, I think

**Comment by borchero** [comment]
@jameslamb the remaining two CI issues are only tangentially related to this PR (namely, `conda` becomes too slow due to the increased complexity of the environment). Do you agree that it would make sense to get this PR fully ready and work on fixing the CI in another PR (e.g. #6140) independently?

File: .ci/test-python-oldest.sh
Comment (jameslamb) [inline .ci/test-python-oldest.sh:12]: Why was this necessary?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Overall, I support the approach to use this vendored-in C API to Arrow dataa instead of taking on a `libarrow` dependency, and on limiting this PR to the Python package + LightGBM C API.

@guolinke @shiyu1994 could you please review these changes, especially those on the C/C++ side?

File: .ci/test-python-oldest.sh
Comment (borchero) [reply .ci/test-python-oldest.sh:12]: This is the constraint that `pyarrow==6.0.1` lists on PyPi.

File: tests/python_package_test/test_arrow.py
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:102]: minor thing, but can you please use the `pytest` fixture `tmpdir` instead of creating one yourself? I think that makes the test code a little easier to understand, and makes it consistent with this project's other tests.

e.g. like this:

https://github.com/microsoft/LightGBM/blob/2d358d5d20e00837ea88455c953c921308c3b69e/tests/python_package_test/test_basic.py#L139

https://github.com/microsoft/LightGBM/blob/2d358d5d20e00837ea88455c953c921308c3b69e/tests/python_package_test/test_basic.py#L162-L163

Otherwise, really like this test! I think it's very thorough way to test that the dataset construction is working as expected.

File: tests/python_package_test/test_arrow.py
Comment (jameslamb) [inline tests/python_package_test/test_arrow.py:12]: As written, this will cause something like `pytest tests/` to fail in an environment where `pyarrow` and `cffi` are not available.

Personally, I'm ok with keeping it that way. That is already how we handle `scikit-learn` tests (another example of tests that rely on a library that isn't a strict requirement for installing `lightgbm`): https://github.com/microsoft/LightGBM/blob/2d358d5d20e00837ea88455c953c921308c3b69e/tests/python_package_test/test_sklearn.py

I'd prefer that this fail with a big loud import error instead of risking a situation where these tests silently don't run in CI.

So not asking for anything to change, just calling that out here for the benefit of @jmoralez and other reviewers.

File: python-package/lightgbm/basic.py
Comment (jameslamb) [inline python-package/lightgbm/basic.py:407]: What's the purpose of having this decorated with `@contextlib.contextmanager` + `yield`-ing? What's the benefit of that over just having a regular function with `return`?

Since this is only going to `yield` exactly once and doesn't have any customized setup (`__enter__()`) or teardown (`__exit__()`) behavior, I don't see the benefit of adding this complexity (e.g. an extra import, a decorator to reason about, a type hint that has to involve `Iterator`).

But I'm also not that familiar with `@contextlib.contextmanager` so I acknowledge there's probably something I'm missing.

File: .ci/test-python-oldest.sh
Comment (jameslamb) [reply .ci/test-python-oldest.sh:12]: I wasn't sure what "on PyPI" meant, so I went looking. Didn't see that in the README at https://pypi.org/project/pyarrow/6.0.1/.

I do see it in the `setup.py` for that version though.

```shell
curl -O https://files.pythonhosted.org/packages/38/c6/97a4133eea642155e7a73cb946d889cadc461a1e6b93f5627af9fdd7b3f3/pyarrow-6.0.1.tar.gz
tar -xvzf ./pyarrow-6.0.1.tar.gz
cat ./pyarrow-6.0.1/setup.py | grep numpy
# 'numpy >= 1.16.6',
```

I think this is acceptable in exchange for testing old versions of `pyarrrow` in this job. `numpy==1.16.6` came out in December 2019 ([link](https://pypi.org/project/numpy/1.16.6/)).

Thanks!

File: tests/python_package_test/test_arrow.py
Comment (jameslamb) [reply tests/python_package_test/test_arrow.py:12]: The alternative would be to add something like this:

https://github.com/microsoft/LightGBM/blob/2d358d5d20e00837ea88455c953c921308c3b69e/tests/python_package_test/test_dask.py#L24-L25

I don't think we should.

**Review by jameslamb — CHANGES_REQUESTED** [review]
Very glad to see the CI is passing!

I'm good with these changes from the perspective of the Python package and testing. I'm not qualified to approve the C API / C++ changes, so let's please wait for @shiyu1994 or @guolinke to take a look.

Marking "request changes" just for the minor new suggestions I left, and to ensure this isn't merged before one of them reviews.

Thanks so much for your work on this!

File: .ci/test-python-oldest.sh
Comment (borchero) [reply .ci/test-python-oldest.sh:12]: > I do see it in the setup.py for that version though.

Yep, that's what I meant :smile: sorry for being imprecise!

File: python-package/lightgbm/basic.py
Comment (borchero) [reply python-package/lightgbm/basic.py:407]: Fair point. My original idea was to make the scope of the C structs clear since `cffi` releases them once the variable referencing them goes out of scope. On second thought, a `with` block doesn't even help with that so let's just use a `return`.

File: tests/python_package_test/test_arrow.py
Comment (borchero) [reply tests/python_package_test/test_arrow.py:102]: Will do! Didn't know that fixture yet :smile:

**Review by guolinke — APPROVED** [review]
The C part looks good to me. Thank you!

**Comment by borchero** [comment]
Thanks for having a look @guolinke!

@jameslamb: I think the CI run can just be re-triggered. Then this should be good to go? 🥳

Once this is merged, I will open (much smaller) PRs to introduce the remaining functionality laid out in #6022. Looking a bit into the future: do you think it is viable to release a new version of LightGBM once this is done?

**Comment by jameslamb** [comment]
> do you think it is viable to release a new version of LightGBM once this is done

Yes, I'd like to do another release soon. But I wouldn't support holding up that PR waiting for this or other Arrow PRs... I think we should try to get the fix to quantized training (#6092) and support for CUDA 12 (#6099) out soon.

**Comment by borchero** [comment]
Ah well, squashing and force-pushing didn't help with the CI @jameslamb 🙄 should I create yet another PR branching off from `master`? :eyes:

**Comment by jameslamb** [comment]
> should I create yet another PR branching off from master? 👀

Yes please, could you? If that doesn't work, I'll go search the GitHub forums.

I really dont want to turn off this setting... this repo has been the target of spam and abuse in the past, and this mechanism helps prevent some specific forms of it.

**Comment by borchero** [comment]
Ahh tried in #6165, no success 😬

> I really dont want to turn off this setting...

Totally understandable. There are two different options though, i.e. whether to require approvals for first-time contributors only or for everyone, right? Are you sure that the correct setting is enabled?

**Comment by shiyu1994** [comment]
@jameslamb  I'm trying to restart the VM from my own laptop with the Linux command line but so far no luck. Perhaps I'll need to restart it from Azure tomorrow. Sorry for the inconvenience.

**Comment by jameslamb** [comment]
Thanks @shiyu1994 , no problem.

**Comment by shiyu1994** [comment]
The machine has been restarted from within Linux. But it seems that the CI jobs are still queued.

**Comment by jameslamb** [comment]
I just tried manually marking all of the other CUDA CI jobs at https://github.com/microsoft/LightGBM/actions/workflows/cuda.yml `failed`, and then restarting them for this PR: https://github.com/microsoft/LightGBM/actions/runs/6699716369

Looks like it's still queued after doing that, like this:

```text
Requested labels: self-hosted, linux
Job defined at: microsoft/LightGBM/.github/workflows/cuda.yml@refs/pull/6034/merge
Waiting for a runner to pick up this job...
```

I can't access the VM so I don't know, but things I'd investigate:

* does the VM have the correct labels in Azure
* is it reachable? (like did the networking configuration change?)
* did something else change in that Azure account, or do we need to re-authorize Azure in this repo's settings somehow?

**Comment by shiyu1994** [comment]
Just restarted the runner service on the VM. And now it works.

**Review by jameslamb — APPROVED** [review]
Thanks for the quick help @shiyu1994 ! Glad they're working again. I'll merge this.

@borchero please update thd other Arrow PRs and I'll provide reviews when I can.

**Pull request merged**
